java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -292,65 +243,83 @@ protected BinaryRelation symmetric() {
   }
 
-  protected abstract RelationState isImpliedBy(BinaryRelation relation);
-
-  protected abstract RelationState impliesEqual();
-
-  protected abstract RelationState impliesNotEqual();
-
-  protected abstract RelationState impliesMethodEquals();
-
-  protected abstract RelationState impliesNotMethodEquals();
-
-  protected abstract RelationState impliesGreaterThan();
-
-  protected abstract RelationState impliesGreaterThanOrEqual();
-
-  protected abstract RelationState impliesLessThan();
-
-  protected abstract RelationState impliesLessThanOrEqual();
-
-  /**
-   * Returns a new relation resulting of the transitive combination of the receiver with the supplied relation.
-   * Schema aRb & bSc => aTc
-   * where aRb is the relation described by the receiver, bSc that of the supplied relation, aTc is the returned relation.
-   * 
-   * @param relation another relation whose left operand is the same as the receiver's right operand
-   * @return the resulting relation or null if the conjunction of both relations does not bring any new information
-   */
-  @CheckForNull
-  protected abstract BinaryRelation combinedAfter(BinaryRelation relation);
-
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithEqual(EqualRelation relation);
+  private Stream<BinaryRelation> combineTransitively(BinaryRelation other) {
+    if (!potentiallyTransitiveWith(other)) {
+      return Stream.empty();
+    }
+    return Stream.concat(combineTransitively0(other), other.combineTransitively0(this));
+  }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithNotEqual(NotEqualRelation relation);
+  private Stream<BinaryRelation> combineTransitively0(BinaryRelation other) {
+    return Stream.of(
+      equalityTransitiveBuilder(other),
+      lessThanTransitiveBuilder(other),
+      greaterThanEqualTransitiveBuilder(other)).flatMap(Function.identity());
+  }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithMethodEquals(MethodEqualsRelation relation);
+  private boolean isEqualityRelation() {
+    return kind == EQUAL || kind == METHOD_EQUALS;
+  }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithNotMethodEquals(NotMethodEqualsRelation relation);
+  private Stream<BinaryRelation> equalityTransitiveBuilder(BinaryRelation other) {
+    if (!isEqualityRelation()) {
+      return Stream.empty();
+    }
+    if (kind == METHOD_EQUALS && other.kind == EQUAL) {
+      return Stream.empty();
+    }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithGreaterThan(GreaterThanRelation relation);
+    return Stream.of(binaryRelation(other.kind,
+      hasOperand(other.leftOp) ? differentOperand(other) : other.leftOp,
+      hasOperand(other.leftOp) ? other.rightOp : differentOperand(other)));
+  }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithGreaterThanOrEqual(GreaterThanOrEqualRelation relation);
+  private Stream<BinaryRelation> lessThanTransitiveBuilder(BinaryRelation other) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)